packages/trie/src/db.ts

```diff
@@ -55,23 +106,79 @@ export class DB {
    * @param value The value to be stored
    */
   async put(key: Buffer, val: Buffer): Promise<void> {
+    const revertOps: BatchDBOp[] = []
+    // In CP mode check for an old value to be put
+    // on the revert ops stack
+    if (this.isCheckpoint) {
+      const oldValue = await this.get(key)
+      if (oldValue !== null) {
+        revertOps.push({
+          type: 'put',
+          key: key,
+          value: oldValue,
+        })
+      }
+    }
     await this._leveldb.put(key, val, ENCODING_OPTS)
+    // In CP mode add del to the revert ops stack
+    if (this.isCheckpoint) {
+      revertOps.push({
+        type: 'del',
+        key: key,
+      })
+      for (const revertOp of revertOps) {
+        this.addCPRevertOperation(revertOp)
+      }
+    }
   }
 
   /**
    * Removes a raw value in the underlying leveldb.
    * @param keys
    */
   async del(key: Buffer): Promise<void> {
+    const revertOps: BatchDBOp[] = []
+    // In CP mode check for an old value to be put
+    // on the revert ops stack
+    if (this.isCheckpoint) {
+      const oldValue = await this.get(key)
+      if (oldValue !== null) {
+        revertOps.push({
+          type: 'put',
+          key: key,
+          value: oldValue,
+        })
+      }
+    }
     await this._leveldb.del(key, ENCODING_OPTS)
+    if (this.isCheckpoint) {
+      for (const revertOp of revertOps) {
+        this.addCPRevertOperation(revertOp)
+      }
+    }
   }
 
   /**
    * Performs a batch operation on db.
    * @param opStack A stack of levelup operations
    */
   async batch(opStack: BatchDBOp[]): Promise<void> {
-    await this._leveldb.batch(opStack, ENCODING_OPTS)
+    if (this.isCheckpoint) {
+      for (const op of opStack) {
+        if (op.type === 'put') {
+          // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+          if (!op.value) {
+            continue
+          }
+          await this.put(op.key, op.value)
+          // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        } else if (op.type === 'del') {
```

Comment:
is this `no-unnecessary-condition` for `if (op.type === 'del')`? is it because it would otherwise just be covered by the `else` clause because type is either `put` or `del`? if so i think i like it more as-is. otherwise could change to `} else {` and just leave a comment below `// type: del`

Reply:
I could remove the first `no-unnecessary-condition` here but linter is complaining on the second. So I will leave half-as-is. 😄